timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/AsyncTimeLockServiceMetadataTest.java

```diff
@@ -0,0 +1,185 @@
+    private final ConjureStartTransactionsRequest startTransactionsRequestWithInitialVersion =
+            ConjureStartTransactionsRequest.builder()
+                    .requestId(UUID.randomUUID())
+                    .lastKnownVersion(ConjureIdentifiedVersion.of(
+                            asyncLockService.leaderTime().id().id(), 0))
+                    .numTransactions(1)
+                    .requestorId(UUID.randomUUID())
+                    .build();
+
+    @Before
+    public void setup() {
+        timeLockService.startWatching(
+                LockWatchRequest.of(ImmutableSet.of(LockWatchReferences.entireTable(WATCHED_TABLE_NAME))));
+    }
+
+    @Test
+    public void absentMetadataIsPassedThrough() {
+        // this will create a lock request for 4 lock descriptors, but with absent metadata
+        IdentifiedLockRequest requestWithoutMetadata = ImmutableIdentifiedLockRequest.copyOf(
+                        standardRequestWithMetadata(ALL_WATCHED_LOCKS_WITH_METADATA))
+                .withMetadata(Optional.empty());
+        assertDone(timeLockService.lock(requestWithoutMetadata));
+
+        assertThat(getAllLockEventsMetadata()).containsExactly(Optional.empty());
+    }
+
+    @Test
+    public void metadataIsPassedThroughForWatchedTable() {
+        assertDone(timeLockService.lock(WATCHED_LOCK_REQUEST_WITH_METADATA));
+
+        assertThat(getAllLockEventsMetadata())
+                .containsExactly(Optional.of(LockRequestMetadata.of(ALL_WATCHED_LOCKS_WITH_METADATA)));
+    }
+
+    // This test is trivial for our current metadata constraints (metadata is only allowed for locks
+    // that are part of the original request), but we would also want this behavior if the constraints weren't there
+    @Test
+    public void noLockEventIsPublishedIfNothingIsWatched() {
+        IdentifiedLockRequest mixedRequest = standardRequestWithMetadata(ImmutableMap.of(
+                UNWATCHED_LOCK_1,
+                ChangeMetadata.updated(
+                        "bla".getBytes(StandardCharsets.UTF_8), "blabla".getBytes(StandardCharsets.UTF_8))));
+        assertDone(timeLockService.lock(mixedRequest));
+
+        assertThat(getAllLockEventsMetadata()).isEmpty();
+    }
+
+    private List<Optional<LockRequestMetadata>> getAllLockEventsMetadata() {
+        return getAllLockWatchEvents().stream()
+                .map(event -> event.accept(LOCK_WATCH_EVENT_VISITOR))
+                .flatMap(Optional::stream)
+                .map(LockEvent::metadata)
+                .collect(Collectors.toList());
+    }
+
+    private List<LockWatchEvent> getAllLockWatchEvents() {
+        ConjureStartTransactionsResponse response =
+                assertDone(timeLockService.startTransactionsWithWatches(startTransactionsRequestWithInitialVersion));
+
+        return response.getLockWatchUpdate().accept(LOCK_WATCH_STATE_UPDATE_VISITOR);
+    }
+
+    private static <T> T assertDone(Future<T> future) {
```

Comment:
I think if we make this method `assert`, we shouldn't have it return a method, outside another type of an `assertion` object.  That said, do we need to call `Future::get` here anyways?

Reply:
We need the future value in `getAllLockWatchEvents`. However, I agree that this method does too much at the same time. I removed it in favor of calling `assertThat(...).isDone()` manually on futures (which is equally short).